### PR TITLE
Add QArray manipulation utilities

### LINKED ISSUE #1080: unitaryHACK - Expand QArray manipulation utilities
### Context

The current `QArray` API lacks standard array manipulation utilities commonly available in NumPy/JAX, such as `swapaxes`, `moveaxis`, `expand_dims`, and `where`.

As a result:

- users often need to convert `QArray` objects to `jax.Array` for basic operations;
- this introduces unnecessary overhead and breaks abstraction consistency;
- the API is less ergonomic and harder to adopt.

This issue aims to close that gap by extending `QArray` with commonly used manipulation primitives.

### Goal

Add support for the following `QArray` manipulation utilities.

**Core additions**

- `swapaxes`
- `moveaxis`
- `expand_dims`
- `where`
- `concatenate`

**Optional additions**
- `take`
- `unstack`

The core additions are required for completion. Optional additions are welcome but not required.

### Expected behavior

The new utilities should follow NumPy/JAX semantics as closely as possible.

They should:

- work directly on `QArray` inputs;
- preserve the `QArray` abstraction in their outputs when appropriate;
- support batching dimensions consistently with the existing `QArray` API;
- avoid requiring users to manually convert to raw `jax.Array`.

### Tests

Add tests comparing the new behavior against the equivalent NumPy/JAX operations.

Tests should cover:

- output shape;
- output values;
- dtype preservation where relevant;
- behavior on batched `QArray` objects;
- compatibility with the existing `QArray` conventions.

### Acceptance criteria

- All core additions are implemented.
- Each core addition has tests.
- Behavior matches NumPy/JAX semantics where applicable.
- Outputs preserve `QArray` behavior where appropriate.
- Existing QArray tests continue to pass.
- Optional additions may be included but are not required.

### Focus

Mostly engineering. Focus on API design, usability, consistency with NumPy/JAX semantics, and integration with the existing `QArray` API.

### Difficulty

Easy

### PR DESCRIPTION
**Pull request by barryPdunham**
Closes #1080

## Summary

Adds QArray manipulation utilities requested in #1080:

* `QArray.swapaxes`
* `QArray.moveaxis`
* `QArray.expand_dims`
* `dq.concatenate`
* `dq.where`

The axis-manipulation methods preserve QArray structure when the operation keeps the final two quantum axes valid, and return a raw JAX array when the operation no longer preserves QArray structure.

`dq.concatenate` supports concatenation along existing batch axes for QArrays with matching dimensions and compatible layouts.

`dq.where` supports dense QArrays directly, preserves sparse DIA layout when the condition is compatible with sparse diagonal storage, and falls back to dense layout with a warning when sparse diagonal structure cannot represent the requested condition.

## Tests

Local validation:

* `ruff check dynamiqs/qarrays tests/qarrays`
* `ruff format --check dynamiqs/qarrays tests/qarrays`
* `ty check dynamiqs`
* `python -m pytest -q tests/qarrays/test_dense_qarray.py tests/qarrays/test_sparse_dia_qarray.py tests/qarrays/test_utils.py`

  * `77 passed, 18 warnings`

The pytest warnings were from running plain pytest locally without the repo’s marker plugin setup, plus an existing sparse-to-dense warning in an existing sparse scalar-add test path.

### CODE REVIEW

**Comment by barryPdunham** [comment]
With the unitaryHACK submission period now complete and #1115 serving as the actively reviewed implementation for #1080, I’m closing this PR as withdrawn.

Thanks for the opportunity to explore the QArray API and contribute an alternative implementation. The work was a useful learning experience, and I appreciate the chance to participate in the project during the hackathon.
